ql/models/shortrate/onefactormodels/blackkarasinski.hpp

```diff
@@ -90,6 +93,39 @@ namespace QuantLib {
         Parameter fitting_;
     };
 
+    class BlackKarasinski::FittingParameter
+        : public TermStructureFittingParameter {
+    private:
+        class Impl :public Parameter::Impl{
+        public:
+            Impl(const Handle<YieldTermStructure>& termStructure,
+                Real a, Real sigma)
+                :termStructure_(termStructure), a_(a), sigma_(sigma) {}
+
+            Real value(const Array&, Time t) const {
+                QL_FAIL("BlackKarasinski::FittingParameter::Impl::value not implemented.");
```

Comment:
If this method throws an exception, the parameter (and therefore the dynamics object that contains it) won't really be usable, am I correct? Calls to their methods will end up calling this one and throwing.

Reply:
Not sure.  Since `phi_` is only used here, though, I'd avoid calibrating it during construction (in most cases, it would be wasted effort) or whenever `generateArguments()` is called.  Maybe we could calculate it on the fly inside `dynamics()` instead of having it as a data member?  I'm open to other suggestions, though.

Reply:
Sure, that makes sense. I will look into it.

Reply:
Hi, I tried something along the lines of our discussion, ```phi_``` is now calibrated inside ```dynamics()```, whenever the latter is called. I also tried to add a test but don't know what is the preferred way of getting a discount factor from the model - I am fairly confident the calibration works fine though.